src/components/Board/Board.js

```diff
@@ -0,0 +1,261 @@
+//add the count (number of adjacent mines arount) to the cell propety
+        let r = props.rows;
+        let c = props.columns;
+        let dir = [[1,0], [-1,0], [0,1], [0,-1], [-1,-1], [-1,1], [1,1], [1,-1]];
+
+        function valid(i,j){
+            if (i>=0 && j>=0 && i < r && j < c){
+                return true;
+            }
+            return false;
+        }
+
+
+        function find(i,j,board){
+            let s = 0;
+            for (let k = 0; k < 8; k ++){
+                let ni = i + dir[k][0];
+                let nj = j + dir[k][1];
+                if (valid (ni,nj))
+                    if (board[ni][nj].hasMine)
+                        s+=1;
+            }
+            return s;
+        }
+
+        
+        function findsomeofneighbors(board){
+
+            for (let i = 0;  i < r; i++){
+                for (let j = 0; j< c; j++)
+                {
+                    board[i][j].count = find(i,j,board)
+                }
+            }
+        }
+
+        findsomeofneighbors(board);
+        return board;
+    };
+
+
+
+
+
+
+
+
+    // a function that gets a cell and incremaens the number of flags and update the boolean of this cel
+    handleLeftClick = Cell =>
+   {
+         Cell.isOpen = true;
+         let newFlags = this.props.flags+ +1 
+         this.setState({flags:newFlags});
+
+        
+   }
+
+        
+
+
+    //a function that opens all the cells around a cell (if has no mines and its not a flag)
+
+   openCleanSpaces = Cell => {
+
+    let currentcell = this.state.table[Cell.y][Cell.x];
+    let newRows = this.state.table;
+    let newOpens = []   //create empty array of all the cells that opend in the first iterations
+    newOpens.push(currentcell);
+    let i = 0;
+
+//iterate over the number of open cells
+for (let i = 0; i<newOpens.length; i++)
+{
+    for (let r = -1; r <= 1; r++)
+    {
+        for (let c = -1; c <= 1; c++)
+          {
+              let newx = newOpens[i].x + c;
+              let newy = newOpens[i].y + r;
+              console.log(newx,newy);
+                if (newx>=0 && newy>=0 && newx < this.state.table.length && newy < this.state.table.length)
+                  {
+                    if (newRows[newy][newx].count === 0 && !newRows[newy][newx].hasMine && !newRows[newy][newx].hasFlag && !newRows[newy][newx].isOpen)
+                    {
+                        newRows[newy][newx].isOpen = true;
+                        newOpens.push(newRows[newy][newx]);
+                        this.props.increamentOpenCells(newOpens.length);
+                  }
+          }
+        
+    }
+
+
+} 
+
+   }
+    this.setState({rows:newRows});
+
+
+
+   }
+
+
+
+
+
+//update the number of flags - when right click
+    updateFlag = Cell => {
+        console.log("here in updateflag function");
+        this.props.increamentFlags();
+        this.props.increamentOpenCells(1);
+        let currentcell = this.state.table[Cell.y][Cell.x];
+        currentcell.hasFlag = true;
```

Comment:
## Make sure not to change existing objects in the state! create new object and replace the old one.
This is a recipe for not-rendering-well components